### PR TITLE
[YDB#11] [NARS] [estess] Speed up $order(gvn,-1) where gvn is of the …

### PR DESCRIPTION
**Pull request by nars1**
…form ^xxx("") and xxx is the global name

Release Note
-------------
An optimization for a specific case makes $order(gvn,-1), where gvn is of the form ^xxx("") and xxx is any global variable name, run faster.

README
-------
See https://github.com/YottaDB/YottaDB/issues/11 for more description of the requirement.

The change involved enhancing gvcst_search to skip gvcst_search_blk() of index blocks for $zprevious(^x(""))
where ^x is any global name (this $zprevious special case is identified by TREF(expand_prev_key) taking on the
special value of ZPREVIOUS_NULL_SUBS_LEVEL1). The consequence is that bh->prev_rec.match and bh->prev_rec.offset
are uninitialized (identified by the macro PREV_REC_UNINITIALIZED) for index blocks after such a gvcst_search() call.
So all places that use prev_rec.match/prev_rec.offset were examined to see if they could be reached using a clue
filled in by a previous $zprevious(^x("")) call. If so, those would have to be examined to do the gvcst_search_blk()
in a deferred fashion. Interestingly, all those places were guaranteed to have prev_rec initialized due to other
reasons. Therefore asserts were added (ASSERT_LEAF_BLK_PREV_REC_INITIALIZED or ASSERT_PREV_REC_INITIALIZED as
appropriate) to indicate prev_rec is initialized wherever it is accessed. A new macro INITIALIZE_PREV_REC_IF_NEEDED
is introduced but never used currently and kept in case it is needed in the future.

Also noticed some unreachable code in gvcst_expand_any_key.c and removed it.

### CODE REVIEW

File: sr_port/gdsfhead.h [hunk 1]
```diff
@@ -3363,6 +3363,58 @@ MBSTART {												\
...
+ * and since sizeof(rec_hdr) is 4 bytes, prev_rec.offset can only be less than 64K-4. And so 64K-1
+ * (i.e. MAXUINT2) serves as a good impossible value for both.
+ */
+#define	PREV_REC_UNINITIALIZED	((unsigned short)MAXUINT2)
+
+/* Note that anytime bh->prev_rec.offset or bh->prev_rec.match is used in the code, one needs to prefix that
+ * usage with either an ASSERT_LEAF_BLK_PREV_REC_INITIALIZED or ASSERT_PREV_REC_INITIALIZED or INITIALIZE_PREV_REC_IF_NEEDED
+ * macro call. This is needed to handle the case where the offset/match could take on the special value PREV_REC_UNINITIALIZED
+ * in case of a special call path through $zprevious(^x("")). If the caller knows for sure it is leaf level block, then it
+ * should use ASSERT_PREV_REC_INITIALIZED. If the caller knows prev_rec is initialized but not guaranteed it is a leaf level
```
Comment (estess) [inline sr_port/gdsfhead.h]: Macro name should be ASSERT_LEAF_BLK_PREV_REC_INITIALIZED instead?

File: sr_unix_cm/rc_prc_getr.c [hunk 2]
```diff
@@ -246,9 +245,8 @@ int rc_prc_getr(rc_q_hdr *qhdr)
 				}
 				rsp->size_remain.value = rc_overflow->size;
 			}
-
-			if (0 == (rc_read_stamp = t_end(&gv_target->hist, two_histories ? &second_history : NULL,
-				TN_NOT_SPECIFIED)))
+			if (0 == (rc_read_stamp = t_end(&gv_target->hist,
```
Comment (estess) [inline sr_unix_cm/rc_prc_getr.c]: Please add comment : (warning assignment)

File: sr_port/gdsfhead.h (hunk 1 above)
Comment (nars1) [inline sr_port/gdsfhead.h]: You are right. Done.

File: sr_port/gdsfhead.h [hunk 3]
```diff
@@ -3363,6 +3363,58 @@ MBSTART {												\
...
+ * block, it should use ASSERT_PREV_REC_INITIALIZED. If neither, then it needs to use INITIALIZE_PREV_REC_IF_NEEDED.
+ * Currently this neither case is not possible (explanations in respective places where one of the above 2 asserts are used).
+ * So INITIALIZE_PREV_REC_IF_NEEDED macro is unused. Nevertheless, its definition is left alone to help when the first need arises.
+ */
+#define	ASSERT_PREV_REC_INITIALIZED(prevRec)			\
```
Comment (nars1) [inline sr_port/gdsfhead.h]: Done.

File: sr_port/gvcst_blk_search.h [hunk 4]
```diff
@@ -84,6 +84,12 @@ GBLREF	uint4			dollar_tlevel;
 #	endif
 #	if defined(GVCST_SEARCH_TAIL) && !defined(GVCST_SEARCH_EXPAND_PREVKEY)
 	assert(0 < memcmp(pKey->base, pOldKey->base, pKey->end + 1));	/* below code assumes this is ensured by caller */
+	/* Before using pStat->prev_rec.offset, we need to make sure it is initialized. i.e. it never holds the value
+	 * PREV_REC_UNINITIALIZED. Thankfully this is guaranteed because "gvcst_search_tail" is currently invoked only for
+	 * leaf blocks. And "gvcst_search" does not currently skip "gvcst_search_blk" for leaf blocks.
+	 * Thankfully the below assert captures all of these so just invoke that.
```
Comment (nars1) [inline sr_port/gvcst_blk_search.h]: Done. :)

File: sr_port/gvcst_search.c [hunk 5]
```diff
@@ -442,12 +459,28 @@ enum cdb_sc 	gvcst_search(gv_key *pKey,		/* Key to search for */
...
 		pCurr->blk_num = nBlkId;
 		pCurr->buffaddr = pBlkBase;
-		if (cdb_sc_normal != (status = gvcst_search_blk(pKey, pCurr)))
-			return status;
-		if (0 == nLevl)
-			break;
-		if ((n0 = pCurr->curr_rec.offset) >= ((blk_hdr_ptr_t)pBlkBase)->bsiz)
-			n0 = pCurr->prev_rec.offset;
+		bsiz = ((blk_hdr_ptr_t)pBlkBase)->bsiz;
+		if (!skip_search_blk || !nLevl)
```
Comment (nars1) [inline sr_port/gvcst_search.c]: Done

File: sr_port/gvcst_search.c [hunk 6]
```diff
@@ -442,12 +459,28 @@ enum cdb_sc 	gvcst_search(gv_key *pKey,		/* Key to search for */
...
+		{
+			assert(PREV_REC_UNINITIALIZED == MAXUINT2);
+			assert(2 == SIZEOF(pCurr->prev_rec.offset));
+			assert(2 == SIZEOF(pCurr->prev_rec.match));
+			pCurr->prev_rec.offset = PREV_REC_UNINITIALIZED;
+			pCurr->prev_rec.match = PREV_REC_UNINITIALIZED;
+			pCurr->curr_rec.offset = bsiz - BSTAR_REC_SIZE;
+			pCurr->curr_rec.match = 0;
+		}
+		if ((n0 = pCurr->curr_rec.offset) >= bsiz)
```
Comment (nars1) [inline sr_port/gvcst_search.c]: Done

File: sr_unix_cm/rc_prc_getr.c [hunk 7]
```diff
@@ -201,23 +204,20 @@ int rc_prc_getr(rc_q_hdr *qhdr)
 						else
 							goto restart;
 				}
-			}
-			else
-				if (fmode & RC_MODE_PREV)	/* if get previous */
+			} else if (fmode & RC_MODE_PREV)	/* if get previous */
+			{
+				if (bh->prev_rec.offset == 0)
```
Comment (nars1) [inline sr_unix_cm/rc_prc_getr.c]: Done

File: sr_unix_cm/rc_prc_getr.c (hunk 2 above)
Comment (nars1) [inline sr_unix_cm/rc_prc_getr.c]: Done

**Pull request merged**